app/src/main/kotlin/com/github/gotify/api/ClientFactory.kt

```diff
@@ -7,57 +7,78 @@ import com.github.gotify.client.api.UserApi
 import com.github.gotify.client.api.VersionApi
 import com.github.gotify.client.auth.ApiKeyAuth
 import com.github.gotify.client.auth.HttpBasicAuth
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import org.tinylog.kotlin.Logger
 
 internal object ClientFactory {
-    private fun unauthorized(baseUrl: String, sslSettings: SSLSettings): ApiClient {
-        return defaultClient(arrayOf(), "$baseUrl/", sslSettings)
+    private fun unauthorized(
+        settings: Settings,
+        sslSettings: SSLSettings,
+        baseUrl: String
+    ): ApiClient {
+        return defaultClient(arrayOf(), settings, sslSettings, baseUrl)
     }
 
     fun basicAuth(
-        baseUrl: String,
+        settings: Settings,
         sslSettings: SSLSettings,
         username: String,
         password: String
     ): ApiClient {
-        val client = defaultClient(
-            arrayOf("basicAuth"),
-            "$baseUrl/",
-            sslSettings
-        )
+        val client = defaultClient(arrayOf("basicAuth"), settings, sslSettings)
         val auth = client.apiAuthorizations["basicAuth"] as HttpBasicAuth
         auth.username = username
         auth.password = password
         return client
     }
 
-    fun clientToken(baseUrl: String, sslSettings: SSLSettings, token: String?): ApiClient {
-        val client = defaultClient(
-            arrayOf("clientTokenHeader"),
-            "$baseUrl/",
-            sslSettings
-        )
+    fun clientToken(settings: Settings, token: String? = settings.token): ApiClient {
+        val client = defaultClient(arrayOf("clientTokenHeader"), settings)
         val tokenAuth = client.apiAuthorizations["clientTokenHeader"] as ApiKeyAuth
         tokenAuth.apiKey = token
         return client
     }
 
-    fun versionApi(baseUrl: String, sslSettings: SSLSettings): VersionApi {
-        return unauthorized(baseUrl, sslSettings).createService(VersionApi::class.java)
+    fun versionApi(
+        settings: Settings,
+        sslSettings: SSLSettings = settings.sslSettings(),
+        baseUrl: String = settings.url
+    ): VersionApi {
+        return unauthorized(settings, sslSettings, baseUrl).createService(VersionApi::class.java)
     }
 
     fun userApiWithToken(settings: Settings): UserApi {
-        return clientToken(settings.url, settings.sslSettings(), settings.token)
-            .createService(UserApi::class.java)
+        return clientToken(settings).createService(UserApi::class.java)
     }
 
     private fun defaultClient(
         authentications: Array<String>,
-        baseUrl: String,
-        sslSettings: SSLSettings
+        settings: Settings,
+        sslSettings: SSLSettings = settings.sslSettings(),
+        baseUrl: String = settings.url
     ): ApiClient {
         val client = ApiClient(authentications)
+        if (settings.legacyCert != null) {
```

Comment:
In https://github.com/gotify/android/pull/344#discussion_r1576873915 you mentioned that some requests fail because multiple clients are created. Can this be fixed with synchronization? I'd expect this happens when two separate threads are trying to migrate the settings and then one sees the legacy cert, tries to migrate it while the other sees no legacy cert and no certPath because it's not completely migrated yet.

How about we add a new method to the Settings class which is executed on all entry points of the app: GotifyApplication and maybe WebSocketService but I'd expect the Application to be called before that.
```kotlin
fun migrate() {
    synchronized(this::class) {
        if (this.legacyCert != null) {
            Logger.info("Migrating legacy CA cert to new location")
            try {
                val legacyCert = this.legacyCert
                this.legacyCert = null
                val caCertFile = File(filesDir, CertUtils.CA_CERT_NAME)
                FileOutputStream(caCertFile).use {
                    it.write(legacyCert?.encodeToByteArray())
                }
                this.caCertPath = caCertFile.absolutePath
                Logger.info("Migration of legacy CA cert succeeded")
            } catch (e: IOException) {
                Logger.error(e, "Migration of legacy CA cert failed")
            }
        }
    }
}
```

Reply:
The next thing I would try to change is the apply() call when changing settings.  
For example legacyCert and caCertPath should be saved synchronously.

Unlike the apply method used [here](https://github.com/cyb3rko/android/blob/client-certificate-auth/app/src/main/kotlin/com/github/gotify/Settings.kt), commit does save content synchronously

Reply:
Yeah, sounds like a good idea. The current solution also doesn't update sslSettings
```kotlin
val client = ApiClient(authentications)
if (settings.legacyCert != null) {
    Logger.info("Migrating legacy CA cert to new location")
    var legacyCert: String? = null
    try {
        legacyCert = settings.legacyCert
        settings.legacyCert = null
        val caCertFile = File(settings.filesDir, CertUtils.CA_CERT_NAME)
        FileOutputStream(caCertFile).use {
            it.write(legacyCert?.encodeToByteArray())
        }
        settings.caCertPath = caCertFile.absolutePath
        Logger.info("Migration of legacy CA cert succeeded")
    } catch (e: IOException) {
        Logger.error(e, "Migration of legacy CA cert failed")
        if (legacyCert != null) settings.legacyCert = legacyCert
    }
}
CertUtils.applySslSettings(client.okBuilder, sslSettings)
client.adapterBuilder.baseUrl("$baseUrl/")
return client
```
`settings.caCertPath` is updated, but the `sslSettings` parameter used in applySslSettings isn't updated. `SslSettings` is immutable.

But I still prefer adding the migration to the GotifyApplication with synchronous writes so the first request should see the updated settings.

Reply:
Works now. And it's moved to the `GotifyApplication`.  
d74bca714c5ce3d534326f0fbdb4763b7d4a6900